### PR TITLE
Fix S3353 FP: Variable is used within an Expression

### LINKED ISSUE #4015: Fix S3353 FP: Variable is used within an Expression
### Description

S3353 is reported for variable if it is used within an `Expression`.
Following the suggestion to make the variable `const` will change the expression.

If the expression is analyzed in frameworks like "NHibernate" the changed expression cause totally different behavior.

I think the rule should not apply for variables using within any `Expression`.

### Repro steps

```cs
namespace SonarDotnetS3353
{
    public class Class1
    {
        public void Method()
        {
            string localVariable = null;

            this.ExpressionMethod(() => localVariable);
        }

        private void ExpressionMethod(Expression<Func<string>> expression)
        {
            if (expression.Body is MemberExpression)
            {
                return;
            }

            throw new InvalidOperationException(
                $"The expression body is a {expression.Body.GetType().Name}, but FieldExpression is expected.");
        }
    }
}
```

When changing `localVariable` to `const string` the analyzing method throws an exception.

### Expected behavior

Rule S3353 should not apply.

### Actual behavior

`SonarDotnetS3353\Class1.cs(10,20,10,41): warning S3353: Add the 'const' modifier to 'localVariable'.`

### Known workarounds

N/A

### Related information

* C#/VB.NET Plugins version: 8.17.0.26580
* Visual Studio version: 2019 16.8.3
* dotnetcore version: 3.1
* Operating System: Windows

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #4015

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs [hunk 1]
```diff
@@ -154,16 +160,35 @@ private static bool HasMutableUsagesInMethod(VariableDeclaratorSyntax variable,
...
+        {
+            if (identifier.FirstAncestorOrSelf<LambdaExpressionSyntax>().GetSelfOrTopParenthesizedExpression() is { } lambda)
+            {
+                if (lambda.Parent is ArgumentSyntax argument
+                    && argument.FirstAncestorOrSelf<InvocationExpressionSyntax>() is { } invocation)
+                {
+                    var lookup = new CSharpMethodParameterLookup(invocation, semanticModel);
+                    return lookup.TryGetSymbol(argument, out var parameter) && parameter.IsType(KnownType.System_Linq_Expressions_Expression_T);
+                }
+                else if (lambda.Parent is AssignmentExpressionSyntax assignment // Lambda cannot be on the left side
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:183]: AFAICT 😃 or can it? 🤔

Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:183]: I think not.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs [hunk 2]
```diff
@@ -154,16 +160,35 @@ private static bool HasMutableUsagesInMethod(VariableDeclaratorSyntax variable,
...
+                    && semanticModel.GetSymbolInfo(assignment.Left).Symbol is { } assignmentTargetSymbol)
+                {
+                    return assignmentTargetSymbol.GetSymbolType().Is(KnownType.System_Linq_Expressions_Expression_T);
+                }
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:187]: We can use `assignment.Left.IsKnownType(KnownType.System_Linq_Expressions_Expression_T, semanticModel)` and simplify the code a bit.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs [hunk 3]
```diff
@@ -352,13 +352,75 @@ public NormalClass NormalClassSetToNull()
...
+
+    public void Noncompliant_Assignment()
+    {
+        string a = null;    // Noncompliant
+        string b = null;    // Noncompliant
+        string c = null;    // Noncompliant
+        var repro = new Repro_4015();
+        Func<string> variableFunc;
+
+        // Compliant cases, changing the Expression<Func<T>> below changes the behavior of the code
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnchangedLocalVariablesShouldBeConst.cs:413]: The cases are marked as compliant but the variables are noncompliant. Did you forgot to remove the comment?

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs:187]: I don't agree on the `a bit`. It's `quite a lot` 😄

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**